### ISSUE TITLE
Added the GlottHMM vocoder and keras

### ISSUE THREAD

**Issue by bajibabu** [CSTR-Edinburgh/merlin#87]
The current merlin scripts are modified to support the GlottHMM vocoder and also added the Keras to train the basic DNN and LSTM models. I have to clean a bit before making a pull-request. In the meantime you can check the modifications here  https://github.com/bajibabu/merlin

**Comment by dreamk73** [comment]
@bajibabu I was able to train a GlottHMM model in merlin yesterday and synthesize it. But I had to change a few paths in the config_default file to get it to work, so I think you need to clean it some more. In one place it says the config_default file is located in the tools/bin/glotthmm directory but then it can't find the pulses directory and the hp_16k file. It also doesn't remove the config_test files if you are not running it from the correct directory.

**Comment by bajibabu** [comment]
Thanks for the feedback. The code needs a bit polishing, that's why I haven't made any PR till now. . Please let us know the quality of GlottHMM for your task.

**Comment by dreamk73** [comment]
Of course. I trained a very small voice yesterday with only 100 sentences. My system is getting parameter files now for a larger set and I will let you know what it sounds like in comparison to WORLD once it is done and trained.

**Comment by dreamk73** [comment]
I trained a female voice with 2400 sentences from our own US English database with both the WORLD vocoder and the GlottHMM vocoder. I didn't tweak any analysis parameters to the voice. I think overall the WORLD vocoder produces a better quality for this voice, but I would like to get rid of the buzziness in the fricatives and /h/ sounds to make it sound even better. 

[world_glotthmm_ex.zip](https://github.com/CSTR-Edinburgh/merlin/files/816424/world_glotthmm_ex.zip)

**Comment by licktion** [comment]
@dreamk73 The quality of the voice you got here is great! So you didn't change any parameter in the codes? Would you like to share the log file during training process, or at least, the model you used and the acoustic model training error and validation error in the end? Thanks a lot!

**Comment by dreamk73** [comment]
Thanks! We use very high quality recordings that we make in-house and the segmentation is manually corrected. The train error is 153.32 and the validation error is 158.71. 
The MCD on the test set is 4.653dB, BAP is 0.184, RMSE is 18.839Hz, VUV is 8.271%.

I am not sure what you mean by I didn't change any parameter in the code? I used six TANH layers of 1024 nodes. I tried with 4 TANH layers and one LSTM layer and it took a lot longer to train, but didn't sound any better. The linguistic features are generated with our own linguistic frontend but it still needs improvement w.r.t. predicting phrasing and accentuation.

**Comment by ronanki** [comment]
@dreamk73 Your VUV is quite high....consider inserting acoustic pauses. A basic phone sequence derived from text may not contain pauses as per the audio file -- so you need to insert pauses where there is a pause in the audio file.

**Comment by dreamk73** [comment]
That is strange. Phonemes and pauses are based on what is found in the acoustic database. The only thing I can think of is that we merge glottal stop labels with the following vowel labels. For this female voice I do notice some buzziness in voiced fricatives that I don't hear nearly as much in our male voice although when I look at the VUV in the male voice it is even higher at 10.781%.

**Comment by ronanki** [comment]
Okay, I forgot to include another reason as well -- inaccurate F0 extraction can also lead to high VUV. 

WORLD does a decent job of F0 extraction -- but can't guarantee for all types of databases.

**Comment by dreamk73** [comment]
That is a good point. I need to do a more detailed analysis of the F0 extraction done by WORLD.

**Comment by dreamk73** [comment]
I retrained my female voice but used code from https://github.com/bjbschmitt/AMFM_decompy to compute F0 values. I learned that you can't use the interpolated F0 values because then the VUV error becomes 0 and there is a lot of buzziness because unvoiced segments are now synthesized as voiced. But when I just used the generated pitch values and retrained Merlin I was able to get slightly better results. I am still hearing some buzzing in places, but it seems to be less frequent and the voiced sections sound better to me. The MCD is similar (4.647 dB vs 4.653), the BAP is similar (0.185 dB vs 0.184), the RMSE is slightly higher (22.164Hz vs 18.839) but the VUV is much better (4.447% vs 8.271%). 

I wonder if there is any tweaking to the WORLD analysis that can be done to improve the pitch calculations for this particular female voice, or whether we should continue with this alternative pitch tracking algorithm.

I have attached a zip with two wavefiles, one with the world F0 and one with AMFM (YAAPT).

[ex_world_amfm.zip](https://github.com/CSTR-Edinburgh/merlin/files/864077/ex_world_amfm.zip)

**Comment by bajibabu** [comment]
Your AMFM samples sounds better than world F0. That is true VUV errors can cause buzzines in synthesized speech.  Can you explain a bit more on what you did here "But when I just used the generated pitch values and retrained Merlin I was able to get slightly better results."

**Comment by dreamk73** [comment]
That is what I thought as well. I am glad to hear that you hear it too ;).

When I first extracted F0 values using the AMFM script, I used their samp_interp values, which smooths the F0 contour to have F0 interpolated in the voiceless sections. I read on here a few days ago that Merlin interpolates values internally anyway, so I figured I would try and see if I could give it a smooth F0 contour. I figured since it has a separate BAP track that it would still not generate voicing in unvoiced sections, but that assumption was incorrect. So instead I took the samp_values from the AMFM pYAAPT algorithm. 

One thing I had to do was write a script to extract these values, make sure I had an equal number of pitch points as WORLD generates and then write it to a binary file in a similar format as WORLD. There is a slight mismatch between the time stamps of WORLD and pYAAPT. WORLD starts at 0.0 ms and then increases every 5 ms. pYAAPT uses half the windowing length as its starting point. I was able to use a 15ms window, so the first time stamp was at 7.5ms. I found that WORLD typically had 3-4 more pitch values than pYAAPT.

**Comment by Ali20177** [comment]
Dear all,

I am just wonder of meaning that the VUV errors = 0%?
In other words, I trained my proposed vocoder parameters under Merlin and I got vuv error = 0%. Is that good? or there is something wrong?

Appreciated your reply.

**Comment by licktion** [comment]
@dreamk73 @bajibabu A quick question: is there any GlottHMM or GlottDNN vocoder available online?

**Comment by ljuvela** [comment]
@licktion  I should have time in April (at last) to finish the GlottDNN documentation and release it.

Here's a mirror for the GlottHMM 1.1 release
https://github.com/mjansche/GlottHMM

**Comment by licktion** [comment]
@ljuvela Thank you so much!

**Comment by licktion** [comment]
@ljuvela @bajibabu @dreamk73 I am trying to use GlottHMM vocoder for Merlin. What I get from feature extraction using GlottHMM vocoder is .F0, .Gain, .HNR, .LSF and .LSFsource binary files. I rename them to .lf0, .gain, .hnr, .lsf and .slsf. I use these five features in the training process, but I got validation error nan, train error nan in fine-tuning model. Any clue?

**Comment by ljuvela** [comment]
By default, GlottHMM 1.1 output format is ASCII, whereas Merlin probably expects 32-bit float. Even if configured to output binary, GlottHMM gives 64-bit float (double). 

You should probably check that the files are in the expected format. SPTK's `x2x` can be used to do this on command line, for example you can do `x2x +af data.ascii > data.float` to convert ascii to 32-bit float. 

Also the .F0 might need special treatment: unvoiced frames are coded with zeros in .F0, so at some point you need separate the voicing into a .vuv file and transform the F0 to log-scale. Finally, interpolate the lf0 over unvoiced frames before training your models. 

It's also possible to use Merlin's default lf0 and vuv, if this seems complicated :)

**Comment by licktion** [comment]
@ljuvela This is really helpful. It's working now. Thank you so much!
